### PR TITLE
Windows: Fix underflow before `delay_usec`

### PR DESCRIPTION
**Pull request by RossRothenstine**
On Windows, a fresh checkout of master with a brand new project was having the CPU sleep for a large amount of time.

Root cause is this math that subtracts target_ticks from the delay resolution, but does not check if target_ticks is less than the delay resolution. The resulting math of: `(target_ticks - delay_resolution) - current_ticks` will then be < 0 and wrap around to max uint64_t. This causes the delay_usec to sleep for around 4 million milliseconds.

### CODE REVIEW

**Comment by clayjohn** [comment]
There might be another issue going on here. In a new project the delay time is 6900, so target ticks will be at least 6900, delay resolution shouldn't be more than 1000

Did you happen to check the value of target ticks and the delay resolution? 

I'm wondering if somehow windows is returning a huge delay resolution

**Comment by RossRothenstine** [comment]
I did record some values here:

current_ticks = 3673038, target_ticks=1, delay_resolution = 1000. (target_ticks - delay_resolution) - current_ticks) = -3674037

This tracks the flow if target_ticks starts at zero and is initialized to current_ticks on the first call of add_frame_delay. dynamic_delay is 1 so it is added to target_ticks of 0. 

current_ticks will be some positive value, however the math of `target_ticks - delay_resolution` will result in an unsigned value of -999 which results in a uint64 underflow.

Compare OS.cpp and OS_Window's logic blocks here, it seems like OS.cpp does not subtract delay resolution, therefore there is no underflow problem.

```
// OS.cpp
	if (dynamic_delay > 0) {
		target_ticks += dynamic_delay;
		uint64_t current_ticks = get_ticks_usec();

		if (current_ticks < target_ticks) {
			delay_usec(target_ticks - current_ticks);
		}

		current_ticks = get_ticks_usec();
		target_ticks = MIN(MAX(target_ticks, current_ticks - dynamic_delay), current_ticks + dynamic_delay);
	}
// OS_Windows.cpp
	if (dynamic_delay > 0) {
		target_ticks += dynamic_delay;
		uint64_t current_ticks = get_ticks_usec();

		// The minimum sleep resolution on windows is 1 ms on most systems.
		if (current_ticks < (target_ticks - delay_resolution)) {
			delay_usec((target_ticks - delay_resolution) - current_ticks);
		}
		// Busy wait for the remainder of time.
		while (get_ticks_usec() < target_ticks) {
			YieldProcessor();
		}

		current_ticks = get_ticks_usec();
		target_ticks = MIN(MAX(target_ticks, current_ticks - dynamic_delay), current_ticks + dynamic_delay);
	}
```

**Review by mrsaturnsan — CHANGES_REQUESTED** [review]
Please use the following. The sleep time should be rounded down to avoid overshooting instead of just subtracting.

**Comment by RossRothenstine** [comment]
> Please use the following. The sleep time should be rounded down to avoid overshooting instead of subtracting.

I'm a bit of a noob but I don't see any code in your comment for a suggestion, mind pasting it in? I'm also not against closing this if a maintainer has a better (less naive) solution. Mine is what unblocked my fork from starting up, but it is a naive fix.

**Comment by RossRothenstine** [comment]
Sounds good, enacted the change and performed a quick smoke test on an empty project, and the project loaded fine.

**Review by mrsaturnsan — APPROVED** [review]
Looks good to me!

**Review by clayjohn — APPROVED** [review]
Makes sense to me. 

I think the situation you are running into comes from the first frame taking longer than the delay time. The old code implicitly assumes that the frame rate always exceeds the target frame rate, but that assumption is not true (especially while loading).

File: platform/windows/os_windows.cpp
Comment (Faless) [inline platform/windows/os_windows.cpp:2265]: Would be nice to leave a comment about why this is being done. Maybe something like:

```
// Make sure we always sleep for a multiple of delay_resolution to avoid overshooting.
// Refer to: https://learn.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-sleep#remarks
```

**Review by Faless — APPROVED** [review]
LGTM, but it would be nice to add a comment explaining the cryptic part where we divide than multiply for the same value.

File: platform/windows/os_windows.cpp
Comment (RossRothenstine) [reply platform/windows/os_windows.cpp:2267]: Clarifying comment added.

**Comment by clayjohn** [comment]
Looks great! The final step before merging is to squash all the commits together so that the whole PR only contains 1 big commit with all your changes. We like to merge one commit at a time to keep the git history clean and navigable. 

If you don't know how to do that, we have a helpful tutorial in the official documentation https://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html#the-interactive-rebase

**Comment by RossRothenstine** [comment]
> Looks great! The final step before merging is to squash all the commits together so that the whole PR only contains 1 big commit with all your changes. We like to merge one commit at a time to keep the git history clean and navigable.
> 
> If you don't know how to do that, we have a helpful tutorial in the official documentation https://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html#the-interactive-rebase

Done and done.

**Pull request merged**
